examples/step-67b/step-67b.cc

```diff
@@ -0,0 +1,1576 @@
+              0.5 * gamma *
+              (pressure_p * (1. / u_p[0]) + pressure_m * (1. / u_m[0]))));
+            const Number s_pos =
+              std::max(Number(), avg_velocity_normal + avg_c);
+            const Number s_neg =
+              std::min(Number(), avg_velocity_normal - avg_c);
+            const Number inverse_s = Number(1.) / (s_pos - s_neg);
+
+            return inverse_s *
+                   ((s_pos * (flux_m * normal) - s_neg * (flux_p * normal)) -
+                    s_pos * s_neg * (u_m - u_p));
+          }
+
+        default:
+          {
+            Assert(false, ExcNotImplemented());
+            return {};
+          }
+      }
+  }
+
+
+
+  template <int dim, typename Number>
+  VectorizedArray<Number>
+  evaluate_function(const Function<dim> &                      function,
+                    const Point<dim, VectorizedArray<Number>> &p_vectorized,
+                    const unsigned int                         component)
+  {
+    VectorizedArray<Number> result;
+    for (unsigned int v = 0; v < VectorizedArray<Number>::size(); ++v)
+      {
+        Point<dim> p;
+        for (unsigned int d = 0; d < dim; ++d)
+          p[d] = p_vectorized[d][v];
+        result[v] = function.value(p, component);
+      }
+    return result;
+  }
+
+
+  template <int dim, typename Number, int n_components = dim + 2>
+  Tensor<1, n_components, VectorizedArray<Number>>
+  evaluate_function(const Function<dim> &                      function,
+                    const Point<dim, VectorizedArray<Number>> &p_vectorized)
+  {
+    AssertDimension(function.n_components, n_components);
+    Tensor<1, n_components, VectorizedArray<Number>> result;
+    for (unsigned int v = 0; v < VectorizedArray<Number>::size(); ++v)
+      {
+        Point<dim> p;
+        for (unsigned int d = 0; d < dim; ++d)
+          p[d] = p_vectorized[d][v];
+        for (unsigned int d = 0; d < n_components; ++d)
+          result[d][v] = function.value(p, d);
+      }
+    return result;
+  }
+
+
+
+  class SubCommunicatorWrapper
+  {
+  public:
+    SubCommunicatorWrapper(
+      const MPI_Comm &   comm,
+      const unsigned int group_size = numbers::invalid_unsigned_int)
+    {
+      if (group_size == 1)
+        {
+          this->comm = MPI_COMM_SELF;
+        }
+      else if (group_size != numbers::invalid_unsigned_int)
+        {
+          const auto rank = Utilities::MPI::this_mpi_process(comm);
+
+          MPI_Comm_split(comm, rank / group_size, rank, &this->comm);
+        }
+      else
+        {
+          const auto rank = Utilities::MPI::this_mpi_process(comm);
+
+          MPI_Comm_split_type(
+            comm, MPI_COMM_TYPE_SHARED, rank, MPI_INFO_NULL, &this->comm);
+        }
+    }
+
+    ~SubCommunicatorWrapper()
+    {
+      if (this->comm != MPI_COMM_SELF)
+        MPI_Comm_free(&comm);
+    }
+
+    const MPI_Comm &get_communicator() const
+    {
+      return comm;
+    }
+
+  private:
+    MPI_Comm comm;
+  };
```

Comment:
This class is new. I am not sure if this is the right place or we want to move something like this into `Utilities::MPI`.

Reply:
I'm in favor of just putting it into a function in `Utilities::MPI`. We already have `Utilities::MPI::duplicate_communicator()`. I don't think you actually need a whole class for this.

Reply:
Maybe more to the point/reason: What you do here currently is conflating two things -- the creation of subcommunicators, and the management of communicators in a C++ fashion. The first happens in the constructor, the second in the constructor/destructor pair providing a lifetime to these communicators. This sort of thing is better handled by separate entities, say a free function (or maybe a static member function) that returns objects of the this kind of type, and then a class that only manages the lifetime of communicators.

MPI used to have C++ wrapper classes for communicators and abandoned them. BOOST also has them, but they have never really gained wide acceptance. deal.II never had wrapper classes for communicators and does not seem to have a great need for them, though one could easily achieve this with `std::unique/shared_ptr` with a custom deleter that just calls `Utilities::MPI::free_communicator()`.

(Actually, on second look, it turns out that we *do* have a communicator wrapper class: `DuplicatedCommunicator`. Except it seems to be completely unused anywhere in the library, and furthermore has the same kind of design flaw as I'm trying to point out above. The person to fault for that is @bangerth ca. 2007.)